use docker compose v2

fmcwalters-edited opened the pull request:
docker compose v2 is part of Docker nowadays. Rather than calling a separate Python lib that calls `docker-compose`, call `docker compose` and use v2.
Fix the `-` and `_` difference between v1 and v2.
This will need THOROUGH testing

PFM-827

## Purpose of PR
Upgrade to use Docker Compose V2
## Parts of the app this will impact
docker compose
## Todos
Test on Linux machines, different CI builds with different packages
## Additional information
This will require Docker with Compose v2
## Related links

fmcwalters-edited committed:
use docker compose v2

docker compose v2 is part of Docker nowadays. Rather than calling
a separate Python lib that calls `docker-compose`, call `docker compose`
and use v2.
Fix the `-` and `_` difference between v1 and v2.
This will need THOROUGH testing

PFM-827

fmcwalters-edited committed:
linting

gchazot commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
 ### Unreleased
 
+# v20.0.0
```
What about `14` ??

fmcwalters-edited replied:
that would make sense 👍 I only put this here as a placeholder so I could test the new version easily. Will change when PR is ready

gchazot commented on kubetools/dev/backends/docker_compose/docker_util.py:
```diff
@@ -137,7 +138,12 @@ def get_containers_status(
-        name = container.name.split('_')[1]
+        print(container.name)
+        print(container.labels.get('kubetools.project.env'))
+        if int(__version__.split('.')[0]) < 20:
+            name = container.name.split('_')[1]
+        else:
+            name = container.name.split('-')[1]
```
Seems like a useless check, this code will only ever be used with this version of the tool, and `__version__` is AFAIK the version of the tool, nothing else

fmcwalters-edited replied:
> I doubt this is going to work as-is. It would be very lucky that all of the features of docker-compose that kubetools uses are fully backwards-compatible in docker compose.

this is why I have stuck this in draft, it still needs a full test across a range of OSes, archs and with as many of our services as possible.

fmcwalters-edited replied:
Also still has print statements! I will remove those alongside the check

fmcwalters-edited replied:
[a439c4c](https://github.com/EDITD/kubetools/pull/163/commits/a439c4cf81449d620986cec9e67d3119f08ba093)

gchazot reviewed (commented):
I doubt this is going to work as-is. It would be very lucky that all of the features of `docker-compose` that `kubetools` uses are fully backwards-compatible in `docker compose`.

fmcwalters-edited commented:
> I doubt this is going to work as-is. It would be very lucky that all of the features of docker-compose that kubetools uses are fully backwards-compatible in docker compose.

This is why it's only in draft, it still needs a thorough test across all dev machine OSes, architectures and with as many packages/services as possible.

fmcwalters-edited commented:
I'm also seeing 
```
WARN[0000] network default: network.external.name is deprecated. Please set network.name with external: true 
```
from running `ktd destroy` - whilst this does not break anything _yet_, I feel like at this stage it is worth fixing.

fmcwalters-edited committed:
use new form of `network` config

the old type, `network.external.name` is deprecated on v2. It still works
but it may be removed in future. Change to new format. See:
https://github.com/docker/compose-cli/issues/1856

fmcwalters-edited committed:
remove version check and `print` statements

this code will only ever be used in this version of the tool, and the
`__version__` only checks the version of the _tool_ anyways.

Also remove redundant print statements

fmcwalters-edited committed:
convert v1 container names to v2

if we create a container using compose v1 and we want to use or remove it
using compose v2, we will need to convert its name to use the new format

fmcwalters-edited committed:
deal with cases where container name has more than one middle name

there are times where the container name consists of multiple names or
a name and then a number; for example, composename-container-name-1-1.
We want to keep only `container-name-1`. Ensure this is possible

fmcwalters-edited commented:
force push because the previous code was weird - it was just to deal with the cases where there was more than one middle name

fmcwalters-edited commented:
we can test this locally using something like `echo -n "cython <3" > /tmp/constraint.txt && PIP_CONSTRAINT=/tmp/constraint.txt pip install -e ~/apps/kubetools` to build and install from this branch. Then ensure you're using v14.0.0 when you run `ktd`.

fmcwalters-edited committed:
use v14.0.0 for Kubetools with Docker Compose v2, update README

Use the most logical version number for the next major version of this.

This version of Kubetools introduces some unavoidable breaking changes
that we can work around. Document these in the README.

fmcwalters-edited committed:
readme formatting

fmcwalters-edited committed:
add unit tests for a complex named app to check compatibility with compose v2

fmcwalters-edited committed:
add missing infra for test

fmcwalters-edited committed:
add correct complex-webserver

fmcwalters-edited committed:
use complex-named-app for cron

fmcwalters-edited committed:
improve documentation
add incompatible version of Docker Desktop to CHANGELOG and add more detail
to README as to why we need to be careful with this update

fmcwalters-edited committed:
ensure dev release adheres to semver format

fmcwalters-edited commented:
I have put this up for a `dev` release - it has been tested to work (with caveats - see `README`) on Mac Intel i5 and Mac M3.
However - we need to test this on Linux systems and in particular on build servers.

fmcwalters-edited committed:
add reason in CHANGELOG why this is a dev release

gchazot commented on README.md:
```diff
@@ -80,15 +80,27 @@ cronjobs:
+of Docker Desktop - [issue here](https://github.com/docker/buildx/issues/2030). 
+There are a couple of workarounds:
+  *  Downgrade to Docker Desktop v4.26.0 or below and append `http://` to each of your `insecure_registries`' URLs
+  *  Alternatively, use the legacy builder by appending `DOCKER_BUILDKIT=0` in front of every `ktd` command.
```
```suggestion
There are a couple of workarounds:
  * Migrate to secure registries (over HTTPS)
  * Downgrade to Docker Desktop v4.26.0 or below and prefix with `http://`, each of your `insecure_registries`' URLs.
  *  Alternatively, use the legacy builder by setting the environment variable `DOCKER_BUILDKIT=0` for `ktd` commands.
```

gchazot commented on README.md:
```diff
@@ -80,15 +80,27 @@ cronjobs:
 
 ## Installing
 
+**NOTE**: before upgrading to version 14.0 or above, you _must_ run `ktd destroy` for all existing 
+local Kubetools projects.
```
I believe this note belongs in CHANGELOG.md because it's a mostly a matter for people upgrading from 13 to 14, and there shouldn't be anything to do special for any new project.

fmcwalters-edited replied:
Have moved this within this commit:
[7a8b920](https://github.com/EDITD/kubetools/pull/163/commits/7a8b920979214f3b3731fce5706bfd3f56f72628)

gchazot commented on README.md:
```diff
@@ -80,15 +80,27 @@ cronjobs:
 
+**NOTE**: before upgrading to version 14.0 or above, you _must_ run `ktd destroy` for all existing 
+local Kubetools projects.
+There is also a bug on the latest version of Docker Desktop that prevents the new Buildkit working with 
```
For the "bug" aspect of it:
* it looks like it's a bug of `buildx` or `moby`, but not "Docker Desktop" which is a distribution of Docker. 
* it would be good to point at a/the ticket that is still open, or fixed, not one that's closed as "not planned"
* this section should be moved out of this section which is about "installing" the tool. It should be moved to a troubleshooting section.
* please add the latest version number of Docker( or moby or buildx) at the time of the PR, and that you have confirmed (tested) to have the bug.

fmcwalters-edited replied:
I have added a better explanation of the bug itself, with a link to the relevant issue in `buildx`, plus a more detailed explanation of what happens if the user is affected by the bug and more detail in the workaround.
I've also cleaned up the CHANGELOG to remove information that shouldn't be there:
[7a8b920](https://github.com/EDITD/kubetools/pull/163/commits/7a8b920979214f3b3731fce5706bfd3f56f72628)

gchazot commented on kubetools/dev/backends/docker_compose/docker_util.py:
```diff
@@ -136,8 +136,14 @@ def get_containers_status(
 
-        # Where the name is compose-name_container_N, get container
-        name = container.name.split('_')[1]
+        # if the container was made in v1, it will be composename_container_N
+        converted_name = container.name.replace('_', '-')
+
+        # we need to keep the middle part of the name
+        # for example if we have a container called `composename-container-1-N`
+        # we want to keep `container-1`
+        middle_names = converted_name.split('-')[1:-1]
+        name = '-'.join(middle_name for middle_name in middle_names)
```
Can you please add tests for this? The main goal being to document the behaviour of this logic and its outcome when applied in each of the possible situations.

Note that the added test "`complex_named_app`" doesn't cover `docker-compose` code. These tests only cover `kubernetes` config generation.

fmcwalters-edited replied:
I've factored this bit of the code out and added tests for different scenarios, plus added test coverage for the `dockerise_label()` function on which this is heavily reliant:
 [b9cc5a4](https://github.com/EDITD/kubetools/pull/163/commits/b9cc5a4cefc1758b48f11241471835c62c3ce99f)

fmcwalters-edited committed:
make workarounds more descriptive

suggest migrating to secure Docker registries, reword the suggestion to use the legacy builder

Co-authored-by: Germain Chazot <germain@edited.com>

fmcwalters-edited committed:
factor out code to convert container names, add unit tests

make the code easier to test by moving the name conversion into its
own function and add test coverage

also add test coverage for `dockerise_label`

fmcwalters-edited commented:
force push only to clean up some mistakes I made with linting and syntax 🤦‍♂️

fmcwalters-edited committed:
improve documentation for buildx issue, make CHANGELOG more concise

add `TROUBLESHOOTING` section to README to document the issue with
`insecure-registries` and link this to ticket which is open and relevant,
for `buildx`.

remove anything from the CHANGELOG which belongs here

gchazot commented on kubetools/dev/backends/docker_compose/config.py:
```diff
@@ -248,9 +248,8 @@ def create_compose_config(kubetools_config):
-                'external': {
-                    'name': 'dev',
-                },
+                'name': 'dev',
```
I'm wondering if we should take this opportunity to rename this "default" network used by `ktd`.

`default` here in the config is from the point of view of the project/container. This declares the default network for the project to be "external" (i.e. pre-existing outside this `compose` project) and named `dev`.

I think we should replace `dev` by something that clearly identifies it as used by `ktd`. Maybe `ktd_dev` could be an option.


Also, from a few tests I've run, it seems like other "environments" as created with `--env` option (or `test` in the case of `ktd test`), end up with a "temporary" network name (i.e. it's not "external" and IS deleted by the destroy operation) which looks like `<appname>_<env>_default`. I think it would be cool if that network name ended up being similarly, something like `ktd_<env>`.
This aspect probably requires a lot of work to implement, However it could be the base for a feature allowing to create "multi-app" environments, in a similar way as we can today do with the "default" environment.

gchazot commented on README.md:
```diff
@@ -137,6 +139,23 @@ minikube delete
+There is a bug in `buildx` which is present in Docker Engine v25.0 and up, which is yet to be patched and causes `insecure-registries` to be ignored - [issue here](https://github.com/docker/buildx/issues/2226).
+Versions of Docker Desktop which use versions of Docker Engine lower than 25.0 are unaffected.
+
+If you try and run `ktd up` with reference to an unsecure registry, e.g. `http://docker-registry.example.net` and are affected by this bug, you will get an error message that is
+similar to the following:
```
```suggestion
If you try and run `ktd up` with reference to an unsecure registry, e.g. `http://docker-registry.example.net` and are affected by this bug, you will get an error message that is similar to the following:
```

gchazot commented on tests/test_docker_util.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class TestDockerComposeNameConversion(TestCase):
+    # we need to check this works for dashes and underscores
```
```suggestion
```

gchazot commented on tests/test_docker_util.py:
```diff
@@ -0,0 +1,37 @@
+from kubetools.dev.backends.docker_compose.docker_util import _get_container_name_from_full_name
+
+
+def generate_long_names(number_of_separators, separator):
```
See comment below, this function probably will disappear. But if it had to stay, then it needs to be renamed because it's confusing::
* It is generating ONE name, not  (several) `_names`
* Why do they have to be "long" ? Is it important for the tests? What would not be tested if the names were "short" ? Note that this aspect of making tests more understandable does not belong in this function's name only, but in how/why the result of it is used in tests.

gchazot commented on tests/test_docker_util.py:
```diff
@@ -0,0 +1,37 @@
+        for i in range(5):
+            long_name = generate_long_names(i, '-')
+            dockerised_name = dockerise_label(long_name)
+            self.assertEqual(dockerised_name, ascii_lowercase[:i+1])
```
I'm not quite sure but I have a feeling that we're using `ascii_lowercase` here as a "long string", which "could look like an app name". If that the case, then in my opinion it's a bad idea. It's a case of "wrong abstraction". Instead here we want to use names that could really be what they stand for, so that it's obvious to the reader what we're talking about. Put them in a list and iterate over them. If you're worried about repeating that list too much, extract it in a constant and reuse it in each place.

gchazot commented on kubetools/dev/backends/docker_compose/docker_util.py:
```diff
@@ -192,6 +191,17 @@ def get_containers_status(
 
 
+def _get_container_name_from_full_name(full_name):
+    # if the container was made in v1, it will be composename_container_N
+    converted_name = full_name.replace('_', '-')
```
Just to check, are we trying to create logic that is backwards compatible with compose V2?

I think that, in line with the change to README and with the major version change, we don´t need to support any backwards compatible logic here.

gchazot commented on tests/test_docker_util.py:
```diff
@@ -0,0 +1,37 @@
+            dockerised_name = dockerise_label(long_name)
+            self.assertEqual(dockerised_name, ascii_lowercase[:i+1])
+
+    def test_container_name_from_full_name_dashes(self):
```
These test names don't explain the scenario that we're testing here.
What are the cases where this logic might be relevant?

fmcwalters-edited committed:
Merge branch 'master' into PFM-827-docker-compose-v2-upgrade

fmcwalters-edited committed:
remove unused import

DilaraOflaz commented on setup.py:
```diff
@@ -59,8 +59,6 @@ def get_readme_content():
             # To support CronJob api versions 'batch/v1beta1' & 'batch/v1'
             'kubernetes>=21.7.0,<25.0.0',
             'tabulate<1',
-            # compose v2 has broken container naming
-            'docker-compose<2',
```
Instead of removing `'docker-compose<2'`, should we add `'docker-compose,'` without specifying a version ?

ryan109 replied:
Docker-compose v2 isn't a python package anymore, I think it's go? Or it's the part of the docker package? I need to refresh my memory on that...

ryan109 replied:
https://docs.docker.com/compose/migrate/

fmcwalters-edited replied:
`compose` is now a subcommand of `docker` itself, and consequently is a Go package that comes with every new Docker installation. There isn't _technically_ `docker-compose` v2, there's `docker-compose` which is a Python package and is now no longer maintained, and `compose` which is a feature of `docker`.

fmcwalters-edited committed:
Merge branch 'master' into PFM-827-docker-compose-v2-upgrade

fmcwalters-edited committed:
use full version number

gchazot reviewed (commented):
Just realised I had all those comments pending from months ago.
